pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -277,40 +296,55 @@ void connectionOpened(ClientCnx cnx) {
 
     }
 
-    void messageReceived(Message message) {
+    void receiveMessagesFromInternalConsumers(int offset) {
+        while (availablePermits.get() > 0) {
+            Message msg = null;
+            // loop around consumers starting from offset consumer.
+            for (int i = offset; i < consumers.size()+offset; i++) {
+                int index = i%consumers.size();
+                try {
+                    msg = consumers.get(index).internalReceive(0, TimeUnit.NANOSECONDS);
+                } catch (PulsarClientException e) {
+                    log.warn("[{}] [{}] Receive listener was interrupted", topic, subscription, e);
+                    break;
+                }
+                if (msg != null) {
+                    // move offset to proceed with next consumer.
+                    offset = index + 1;
```

Comment:
If consumer at `offset` has messages and we don't advance `offset` we would always be pulling from the same consumer in each iteration, this would be unfair for the rest.
By advancing we can at least try the other consumers, eventually returning to the same if no one had